### DISCUSSION TITLE
How to make a property reactive with attribute but not need to trigger component render.

### DISCUSSION THREAD

**Discussion by XGHeaven** [lit/lit#2163] [Q&A]
For example. There has a component with an attribute named `theme`. `theme` attribute is only used in css selector for the theme and click event.

```typescript
class SomeElement extends LitElement {
  static styles = css`
  :host { color: black; }
  :host([theme=dark]) { color: white; }
  `

  @property({type: String, reflect: true, /* which option should write here */}) theme?: string

  render() {
    // in render function , no using `theme` property
    return html`<div><slot></div>`
  }

  firstUpdated() {
    this.addEventListener('click', e => {
      if (this.theme === 'dark') {
        // do something with a dark theme
      }
      // do something with a normal theme
    })
  }
}
```

If I update theme property or attribute. The component always is trigger render. This is not what I expected.
I cannot found an option to prevent this.

I tried to use getter/setter instead of @property decorator. It's works but so troublesome.

**Comment by kevinpschaaf** [comment]
This is a pretty uncommon case so we don't have an option for this exact use case, but here are a few options to consider:

1. We have a `noAccessor: true` option which won't create an accessor that triggers reactive updates when the property/attribute is set (caveat known issue below), however `reflect: true` is implemented as part of the reactive lifecycle, so turning that off means you could not set the property and have the attribute-based styling continue to work.

1. If you're ok just treating `theme` as an attribute-only API, you can skip making an `@property` altogether and just read `this.getAttribute('theme')` instead of `this.theme` in your click event listener.

1. You can implement `shouldUpdate` and check `changedProperties` and return `false` if it only includes `theme`. If this is a regular pattern used in a lot of components, you could abstract that to a base class.

However, can you expand on why it is important for `theme` to _not_ trigger a render? In general, `render` should be idempotent w.r.t. element state, so unless you're hyper-optimizing for a high-frequency theme change, the no-op render should be unobservable other than (low-cost) work of doing the no-op render.

> *Note there is currently a developer-mode bug affecting `noAccessor` https://github.com/lit/lit/pull/2160 that will be resolved in the next release candidate.

**Reply by XGHeaven** [reply]
> 3. You can implement `shouldUpdate` and check `changedProperties` and return `false` if it only includes `theme`. If this is a regular pattern used in a lot of components, you could abstract that to a base class.

It can't synchronize property to attribute if I return `false` of `shouldUpdate` function. I tried.

> However, can you expand on why it is important for theme to not trigger a render?

Some attribute is not impact whats the DOM looks like but needs to be read as property or keep synchronous between property and attribute. For example, In web spec, `title`/`required` property in `<input/>`/`<textarea/>` DOM. Only attribute change when property change. DOM is not changed.

```typescript
const input = document.createElement('input')
// <input/>

input.title = 'abc'
// <input title="abc/>
```

This is why it's so important. And I think this is a common case for webcomponent. And many built-in HTML tags also depends on this.

**Comment by kevinpschaaf** [comment]
> Only attribute change when property change. DOM is not changed.

If the changed property is not used in `render`, the DOM _will not be changed_. When Lit re-runs a `render`, although it may look like Lit is re-concatenating the strings and setting the HTML, this is not how Lit actually works. On a re-render, only the _values_ (i.e. the result of the expressions in each `${...}` are compared to the previous values, and only when they differ is the DOM changed.

Based on that understanding, can you try and answer this question again?

> However, can you expand on why it is important for theme to not trigger a render? In general, render should be idempotent w.r.t. element state, so unless you're hyper-optimizing for a high-frequency theme change, the no-op render should be unobservable other than (low-cost) work of doing the no-op render.

Lit is designed to use a very simple mechanism for triggering the update lifecycle: It creates accessors for each `@property`, and when a property is set to a different value, it triggers an update. The update lifecycle is responsible for running several callbacks, including `update()` (which implements `reflect: true` attribute reflection) and passing the results from `render()` to `lit-html`, which will update any changed values (or no-op if they are all the same as previous).

We currently don't have a way to separate the part of the update lifecycle that does attribute reflection from the part of the update lifecycle that does render based on individual property options, since that would introduce extra complexity for a level of optimization is rarely necessary.
